### PR TITLE
CB-16723 Capture telemetry metrics for newly introduced flows in Stop…

### PR DESCRIPTION
**Pull request by aakashshankar**
…Start scaling

- Added the new flows to ClusterUseCaseMapper to propagate metrics for faster autoscaling. 

See detailed description in the commit message.

### CODE REVIEW

File: structuredevent-service-legacy/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/telemetry/mapper/ClusterUseCaseMapper.java [hunk 1]
```diff
@@ -98,6 +102,10 @@ private UsageProto.CDPClusterStatus.Value lastStepToUseCaseMapping(String rootFl
                     useCase = getClusterStatus(nextFlowState, "FLOWCHAIN_FINALIZE_FINISHED_STATE",
                             UsageProto.CDPClusterStatus.Value.DOWNSCALE_FINISHED, UsageProto.CDPClusterStatus.Value.DOWNSCALE_FAILED);
                     break;
+                case "StopStartUpscaleFlowEventChainFactory":
+                    useCase = getClusterStatus(nextFlowState, "STOPSTART_UPSCALE_FINALIZE_STATE",
+                            UsageProto.CDPClusterStatus.Value.STOP_START_DOWNSCALE_FINISHED, UsageProto.CDPClusterStatus.Value.STOP_START_DOWNSCALE_FAILED);
```
Comment (sidseth) [inline structuredevent-service-legacy/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/telemetry/mapper/ClusterUseCaseMapper.java:107]: This should be the UPSCALE protos?

And, StopStartDownscaleEventChainFactory missing ?

**Review by sidseth — COMMENTED** [review]
Needs changes. That said, I believe this set of changes is adequate after the change request is in place.

One addition: I think it'll be useful to add tests to make sure the States referenced as string literals do not get renamed - along with a comment indicating the intent of the test. Also, comments in StopStart*FlowConfig, and StopStart*FlowConfig indicating that they should not be renamed, since metrics depends on the string literal (or a test to verify the class is not renamed if that is possible... would be quite a strange test).

Will let @Bajzathd , @nadamdb review.

File: structuredevent-service-legacy/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/telemetry/mapper/ClusterUseCaseMapper.java (hunk 1 above)
Comment (aakashshankar) [reply structuredevent-service-legacy/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/telemetry/mapper/ClusterUseCaseMapper.java:111]: Sorry about this, misinterpreted the method. Updated the protos and added the `StopStartDownscaleFlowEventChainFactory`

**Comment by aakashshankar** [comment]
>  I think it'll be useful to add tests to make sure the States referenced as string literals do not get renamed - along with a comment indicating the intent of the test. Also, comments in StopStartFlowConfig, and StopStartFlowConfig indicating that they should not be renamed, since metrics depends on the string literal (or a test to verify the class is not renamed if that is possible... would be quite a strange test).

Went through the current UTs for this and it looks like the telemetry useCase is automatically set to `UNSET` for unrecognized string literals (FlowStates). @nadamdb @Bajzathd is there any way to make sure that these literals (or the Flow state) is not renamed so that metrics are always propagated?

**Comment by aakashshankar** [comment]
retest this please

**Review by nadamdb — APPROVED** [review]
LGTM

**Comment by nadamdb** [comment]
> > I think it'll be useful to add tests to make sure the States referenced as string literals do not get renamed - along with a comment indicating the intent of the test. Also, comments in StopStartFlowConfig, and StopStartFlowConfig indicating that they should not be renamed, since metrics depends on the string literal (or a test to verify the class is not renamed if that is possible... would be quite a strange test).
> 
> Went through the current UTs for this and it looks like the telemetry useCase is automatically set to `UNSET` for unrecognized string literals (FlowStates). @nadamdb @Bajzathd is there any way to make sure that these literals (or the Flow state) is not renamed so that metrics are always propagated?

@aakashshankar I'm not aware of any mechanics that cloud prevent the flows/flowsteps/flowchains not to be renamed. Last time I checked the flowcahin names are not used anywhere else besides this class (not even exposed to UI or something) and by default the class name is the flowchain name and it's not likely that it's overridden. Same for the flow names, I don't think that they are frequently renamed. One possible issue could be however if the flowchain is reconfigured and the starting/ending flows are changed to something else, then the mapper should be also updated to follow the change.

**Comment by sidseth** [comment]
> > I think it'll be useful to add tests to make sure the States referenced as string literals do not get renamed - along with a comment indicating the intent of the test. Also, comments in StopStartFlowConfig, and StopStartFlowConfig indicating that they should not be renamed, since metrics depends on the string literal (or a test to verify the class is not renamed if that is possible... would be quite a strange test).
> 
> Went through the current UTs for this and it looks like the telemetry useCase is automatically set to `UNSET` for unrecognized string literals (FlowStates). @nadamdb @Bajzathd is there any way to make sure that these literals (or the Flow state) is not renamed so that metrics are always propagated?

Can we add a simple unit test under the StopStart* module ? All it needs to do is to compare the names of the states, and classes using strings.
e.g.
Asseert.assertEquals("com.sequenceiq.cloudbreak.core.flow2.chain.StopStartDownscaleFlowEventChainFactory", StopStartDownscaleFlowEventChainFactory.class.getName(), "ClassName not as expected. This may cause issues with EDH metrics propagation. Ref blah");

4-5 1 line tests (on arguably the wrong module) helps prevent accidental errors. (More relevant here - since the 'StopStart' name may actually change)

**Comment by horadla23** [comment]
@sidseth @nadamdb 
please discuss this matter within a separate ticket, since I would guess this logic to ensure string literal correctness applies for every metrics, thus a general enforce logic would be useful, to ensure noone will rename these literals.

**Pull request merged**